cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -199,249 +214,223 @@ __global__ void subtract_set_bits_range_boundaries_kernel(bitmask_type const* bi
-                                                  IndexIterator indices_end,
-                                                  rmm::cuda_stream_view stream)
+std::vector<size_type> segmented_count_bits(bitmask_type const* bitmask,
+                                            IndexIterator indices_begin,
+                                            IndexIterator indices_end,
+                                            count_bits_policy count_bits,
+                                            rmm::cuda_stream_view stream)
 {
-  size_t const num_indices = std::distance(indices_begin, indices_end);
+  auto const num_indices = static_cast<size_type>(std::distance(indices_begin, indices_end));
+  CUDF_EXPECTS(num_indices % 2 == 0, "Array of indices needs to have an even number of elements.");
+  size_type const num_ranges = num_indices / 2;
+
+  for (size_type i = 0; i < num_ranges; i++) {
+    auto begin = indices_begin[i * 2];
+    auto end   = indices_begin[i * 2 + 1];
+    CUDF_EXPECTS(begin >= 0, "Starting index cannot be negative.");
+    CUDF_EXPECTS(end >= begin, "End index cannot be smaller than the starting index.");
+  }
 
   if (num_indices == 0) {
     return std::vector<size_type>{};
   } else if (bitmask == nullptr) {
-    return std::vector<size_type>(num_indices / 2, 0);
+    std::vector<size_type> ret(num_ranges, 0);
+    if (count_bits == count_bits_policy::SET_BITS) {
+      for (size_type i = 0; i < num_ranges; i++) {
+        ret[i] = indices_begin[2 * i + 1] - indices_begin[2 * i];
+      }
+    }
+    return ret;
   }
 
-  auto ret = segmented_count_set_bits(bitmask, indices_begin, indices_end, stream);
-  for (size_t i = 0; i < ret.size(); i++) {
-    auto begin = indices_begin[i * 2];
-    auto end   = indices_begin[i * 2 + 1];
-    ret[i]     = (end - begin) - ret[i];
-  }
+  // Construct a contiguous host buffer of indices and copy to device.
+  auto const h_indices = std::vector<size_type>(indices_begin, indices_end);
+  auto const d_indices = make_device_uvector_async(h_indices, stream);
+
+  // Compute the null counts over each segment.
+  auto first_bit_indices = thrust::make_transform_iterator(
+    thrust::make_counting_iterator(0), index_alternator{false, d_indices.data()});
+  auto last_bit_indices = thrust::make_transform_iterator(thrust::make_counting_iterator(0),
+                                                          index_alternator{true, d_indices.data()});
+  rmm::device_uvector<size_type> d_bit_counts = cudf::detail::segmented_count_bits_device(
+    bitmask, num_ranges, first_bit_indices, last_bit_indices, count_bits, stream);
+
+  // Copy the results back to the host.
+  std::vector<size_type> ret(num_ranges);
+  CUDA_TRY(cudaMemcpyAsync(ret.data(),
+                           d_bit_counts.data(),
+                           num_ranges * sizeof(size_type),
+                           cudaMemcpyDeviceToHost,
+                           stream.value()));
```

Comment:
Do we not have a factory for this already? If not, let's make one. Raw `cudaMemcpyAsync` calls are usually an indicator that we're missing some functionality.

Reply:
I see this pattern repeated a lot of times in the codebase, when [searching for `CUDA_TRY cudaMemcpyAsync`](https://github.com/rapidsai/cudf/search?p=5&q=CUDA_TRY+cudaMemcpyAsync). We have [`hostdevice_vector` with a couple helper methods for copying between host and device](https://github.com/rapidsai/cudf/blob/d1811b5baf1d83f8d376a4f6e7fd84020a24506b/cpp/src/io/utilities/hostdevice_vector.hpp#L105-L117). However, that class isn't well suited for use cases like this one where the data is already owned by the device.

Reply:
I found the function I need: `make_std_vector_async`.

https://github.com/rapidsai/cudf/blob/1697f63b9e6e80695cb157f479fada72d053fa1a/cpp/include/cudf/detail/utilities/vector_factories.hpp#L328-L329

This would make a great follow-up PR for fixing the other uses of `CUDA_TRY(cudaMemcpyAsync(...))`.

Reply:
I have created #9806 to follow up on this suggestion.